packages/java-edition/src/mcfunction/parser/argument.ts

```diff
@@ -1640,3 +1731,173 @@ function vector(
 		return ans
 	}
 }
+
+function components(): core.InfallibleParser<ComponentListNode> {
+	return core.map(
+		core.record({
+			start: '[',
+			pair: {
+				// TODO: Create an item_component category
+				key: core.resourceLocation({
+					pool: [],
+					allowUnknown: true,
+					namespacePathSep: '.',
+				}),
+				sep: '=',
+				value: nbt.parser.entry,
+				end: ',',
+				trailingEnd: true,
+			},
+			end: ']',
+		}),
+		(res) => {
+			const ans: ComponentListNode = {
+				type: 'mcfunction:component_list',
+				range: res.range,
+				children: res.children,
+			}
+			return ans
+		},
+	)
+}
+
+function componentTests(): core.InfallibleParser<ComponentPredicatesNode> {
+	const exact = core.map<
+		core.SequenceUtil<
+			core.ResourceLocationNode | core.LiteralNode | nbt.NbtNode
+		>,
+		ComponentTestExactNode
+	>(
+		core.sequence([
+			core.resourceLocation({
+				pool: [],
+				allowUnknown: true,
+				namespacePathSep: '.',
+			}),
+			core.literal('='),
+			nbt.parser.entry,
+		]),
+		(res) => {
+			const id = res.children.find(core.ResourceLocationNode.is)!
+			const value = res.children.find(nbt.NbtNode.is)!
+			const ans: ComponentTestExactNode = {
+				type: 'mcfunction:component_test_exact',
+				range: res.range,
+				children: [id, value],
+				component: id,
+				value,
+				negated: false,
+			}
+
+			return ans
+		},
+	)
+
+	const exists = core.map<
+		core.ResourceLocationNode,
+		ComponentTestExistsNode
+	>(
+		core.resourceLocation({
+			pool: [],
+			allowUnknown: true,
+			namespacePathSep: '.',
+		}),
+		(res) => {
+			const ans: ComponentTestExistsNode = {
+				type: 'mcfunction:component_test_exists',
+				range: res.range,
+				children: [res],
+				component: res,
+				negated: false,
+			}
+
+			return ans
+		},
+	)
+
+	const subpredicate = core.map<
+		core.SequenceUtil<
+			core.ResourceLocationNode | core.LiteralNode | nbt.NbtNode
+		>,
+		ComponentTestSubpredicateNode
+	>(
+		core.sequence([
+			core.resourceLocation({
+				pool: [],
+				allowUnknown: true,
```

Comment:
This should use the `item_sub_predicate_type` registry (already exists in mcmeta registries, just needs to add to `RegistryCategories`).